### DISCUSSION TITLE
Example with many character mappings

### DISCUSSION THREAD

**Discussion by bitoj** [cajhin/capsicain#21] [Show and tell]
For comparison an perhaps inspiration, here is the core of my current configuration. It is work in progress, with little testing after creating it based on a prevous AutoHotkey script. There might still be some typos, cut/paste errors or unintended results of my conversion script.

I used the folowing design criteria.
- Reduce strain when typing in Dutch (without the stretching required with default US International Windows keyboard lay-out for NL) or Esperanto.
- ANSI or US ISO as the main hardware lay-outs.
- Provide key combinations for several typographic quote styles and dashes.
- Provide access to key codes for keys that are not physically present on my ergonomic keyboard, but are required in my work environment.
- Pass through everything to a remote work computer, where I cannot install special supporting software. (This is the main reason why I switched from AHK.)
- Provide predictable access, even if cumbersome, to all characters to enter European _names_ in Latin, Greek and Cyrillic scripts. (I am not a polyglot and entering long stretches of foreign language text is not my purpose.)

Notable features are, I think:
- extensive use of second order deadkeys, i.e., using deadkeys to define deadkeys;
- starting applications, even remotely, using key sequences with `LWIN+R` and executable name or path;
- limited use of `REWIRE` with all parameters, but no `//`.

For common text editing I use `CAPS` and for digits `TAB`, but somewhat differently than cajhin. I am not a gamer, so WASD or IJKL are not natural to me, but the HJKL keys from `vi` are. Other navigation and editing key combinations with `CAPS` are next to that row.

The following fragment illustrates second order deadkey definitions.
```
COMBO LSHF [...& ....] > deadkey(RSHF)

# grave accent
COMBO DEADKEY-RSHF ` [^^^^ ^^^^] > deadkey(`)
COMBO DEADKEY-`    A [...^ ...^] > altChar(0224)  # à
COMBO DEADKEY-`    A [...| ...|] > altChar(0192)  # À

# acute accent and several single quotes
COMBO DEADKEY-RSHF ' [^^^^ ^^^^] > deadkey(')
COMBO DEADKEY-'    E [...^ ...^] > altChar(0233)  # é
COMBO DEADKEY-'    E [...| ...|] > altChar(0201)  # É
COMBO DEADKEY-'    [ [...^ ...^] > altChar(0145)  # ‘
COMBO DEADKEY-'    / [...^ ...^] > altChar(0146)  # ’
COMBO DEADKEY-'    , [...^ ...^] > altChar(0130)  # ‚
COMBO DEADKEY-'    , [...| ...|] > altChar(0139)  # ‹
COMBO DEADKEY-'    . [...| ...|] > altChar(0155)  # ›
COMBO DEADKEY-'    \ [...^ ...^] > altChar(08219) # ‛
```
See [capsicain.ini.txt](https://github.com/cajhin/capsicain/files/5824477/capsicain.ini.txt).

**Comment by cajhin** [comment]
Whoa there. That's a serious config.
And that's a wild amount of special characters. I hope you have a system in your head, because memorizing all of that is rather impossible.

Side note: this made me imagine a neat feature, where deadkeys trigger a little overlay in the corner that show the possible combinations.
Like > (deadkey ~) triggers
```
┌─────┐
│ n ñ │
│ a ã │
│ o õ │
└─────┘
```
Your config looks really nice and clean with only one thing I noticed:
Rewire press-and-Hold:

`REWIRE RALT MOD11 MOD11 RALT`

What happens when you press Alt, release, press again?
1. RALT down -> MOD11 down
2. RALT up   -> MOD11 up
3. 'iftapped' triggered -> MOD11 down and up
4. 'tap&Hold' triggered -> RALT down
Step 3 is useless. That should be ignored. So your Rewire should be:

`REWIRE RALT MOD11 // RALT`

You're not using MOD11 though. You should, I do it all the time, and it's awesome
(especially with Mac boards or similar, that have a smaller space bar).


Starting apps remotely by sending Win+R sequences is clever, never thought of that.


**Details about your layout, where I prefer mine**
This ain't a pissing contest; the best config is always one you're comfortable with, though, and it sounds like you already learned the config on AHK

caps mod9 caps: my caps is probably my most used key after space, and single purpose (except for tapping combos). When I feel an urge to edit, I automatically press caps. If I reconsider, I can release it again.
With your config you have to be careful.

apps mod12: Apps is a super useless key and invites you to take it over. But your fkd when you get a board that omits this useless key...
A lot of nice smaller boards don't have it.

caps 4 = €: bit difficult to press. I prefer my "tapped alt + E" (I just noticed that my logitech even has an € symbol on E)

caps tab [] = () bsp del "because less stretching"
I love my caps H = backspace. Very frequent key, on the homerow next to my right index finger, and somehow intuitive.
Also, as a programmer, I type (){}[] all day, every day. Your config puts all that on the extended right pinky, which is my worst finger (from hitting Return I guess).
I love to have the brackets on the strong ring/middle fingers, with Alt iokl,. = (){}[]
And I can type (){}[] in half a second. That's a coder thing, though...

caps p/n = word left/right
doesn't make sense to me. "Left" is my index finger's job. Is that a vi thing?

Your left shift is now nice. When you get used to it, you'll want symmetry for your right shift.
I put that on the /?-_ key. Now you need a new home for <>|/?...
With your Alt->MOD11 rewire, you can put ALL the 'regular' special characters on the A-Z keys and never again stretch for shift-6.
This is relatively new in my setup, took a few days to become comfortable, but now I really really love it.

**Reply by bitoj** [reply]
Thanks. You gave me several pointers for improving my configuration and make the best of the new possibilities that you also provided with capsicain.

It is especially interesting to see a different perspective. I am not really a coder (wrote my last line of C++ about 27 years ago), but your examples made me realize I need to reevaluate my current usage patterns. My configuration has a long history with inspiration from vi and XCompose. It also still shows some effect of earlier approaches with MS KLC and AHK and the limitations of those programs. As you noticed, it is also rather unbalanced, because for a long time I had to combine it with AltGr patterns of the US International Windows keyboard layout that I was forced to use in a corporate environment.

I think now is the opportunity for a fresh start, at least for all navigation and editing commands and input in my native language. I think I'll keep the deadkey definitions, because they are about exceptions, that primarily need a system that I can remember, not efficiency.

The coming weeks I'll have a closer look at your configuration and see which parts are appropriate for me and where I have to make adjustments. I'll probably start with shuffling my modifiers, with your recommendations as a starting point.
